### ISSUE TITLE
"Illegal instruction" crash on Apple Silicon

### ISSUE THREAD

**Issue by nathunsmitty** [sorbet/sorbet#3852]
#### Input

The sorbet-static binary crashes on new M1 arm-based macs with the error `Illegal instruction`.

```
srb typecheck -e 'puts "Hello, world!"'
```

Attempting to run Sorbet under Rosetta 2 also fails.

```
arch -x86_64 srb typecheck -e 'puts "Hello, world!"'
```

#### Observed output

Both with and without Rosetta 2:

```
/usr/local/lib/ruby/gems/2.6.0/gems/sorbet-0.5.6189/bin/srb: line 46: 37685 Illegal instruction: 4  "${sorbet}" "${args[@]}"
```

#### Expected behavior

Sorbet correctly typechecks the given statement by printing `No errors! Great job.`

#### My Research

I dug into this issue a bit and came across the very similar #1487. (Apologies if this is a dupe.) @DarkDimius' [suggestion](https://github.com/sorbet/sorbet/issues/1487#issuecomment-521506401) of removing the `-march=sandybridge` line produced a binary that appears to be working on my machine for `--config=dbg`. `--config=release-mac` caused compile issues with some libraries.

Since I'm assuming you don't want to give up fancy Intel optimizations, it seems like the best solution would be to [create a universal binary](https://developer.apple.com/documentation/xcode/building_a_universal_macos_binary) through merging an ARM build and an x86 build via `lipo` or another tool. (This is alluded to in #3369.) 

Cross-compiling C++ binaries is a little outside my area of expertise, but I'm happy to help in any way I can!

**Comment by dnsco** [comment]
Hey, I'm trying to figure out how to work with this situation,

looks like I've gotta download a preview build of open jdk and patch bazel to build myslef, I was wondering if anyone has come up with an easier solution?

**Comment by nathunsmitty** [comment]
Unfortunately not. 🙁 

Here's my set up:

1. A custom fork of Sorbet. (You can make your own, or use [mine](https://github.com/nathunsmitty/sorbet)!) There are two lines in `.bazelrc` that I changed:
    1. [Remove the `-march=sandybridge` argument.](https://github.com/nathunsmitty/sorbet/commit/c71789e31ac9a8b147bdbc87e683adb099c47cfe)
    2. [Don't include jemalloc on the `mac-release`.](https://github.com/nathunsmitty/sorbet/commit/c0e415f7676fd04f6e175eabdaf786b89e94c752) I ran into a lot of clang errors with jemalloc due to changes in the default error warnings on Big Sur. There's probably a way to pass the right settings to Bazel when it builds jemalloc but I couldn't figure it out.
2. Build Sorbet with `./bazel build //main:sorbet --config= release-mac `
3. The M1 compatible executable is at `bazel-bin/main/sorbet`!
4. I'm using Sorbet with a Rails app, so I just copied the executable to my `<project>/bin` directory. I run Sorbet with `bin/srb`.
5. Update your editor LSP paths accordingly.

This is obviously pretty hacky, but it works for me and my admittedly small project. If other people can improve on this process (particularly the jemalloc part), I'd love to hear about it!

**Comment by connorshea** [comment]
@nathunsmitty I realize I can compile it myself, but would you be willing to publish your custom build to rubygems for others to use?

**Comment by nathunsmitty** [comment]
@connorshea done: https://rubygems.org/gems/sorbet-static-m1. Sorry it took so long, it's been a busy week. Let me know if you run into any issues!

**Comment by connorshea** [comment]
@nathunsmitty thanks! No worries on the wait, my M1 Mac is still in transit so I won't have it until next week anyway :P

**Comment by jez** [comment]
Closed in #4051
